hoodie-common/src/main/java/com/uber/hoodie/common/util/SerializationUtils.java

```diff
@@ -16,156 +16,126 @@
-      throw new HoodieSerializationException("cannot cast class", ex);
-    } catch (final ClassNotFoundException ex) {
-      throw new HoodieSerializationException("class not found", ex);
-    } catch (final IOException ex) {
-      throw new HoodieSerializationException("unable to deserialize to object", ex);
+      return (T) serializerInstance.deserialize(objectData);
     } finally {
-      try {
-        if (in != null) {
-          in.close();
-        }
-      } catch (final IOException ex) { // NOPMD
-        // ignore close exception
-      }
+      releaseSerializerInstance(serializerInstance);
     }
   }
 
-  /**
-   * <p>
-   * Deserializes a single {@code Object} from an array of bytes.
-   * </p>
-   *
-   * <p>
-   * If the call site incorrectly types the return value, a {@link ClassCastException} is thrown from the call site.
-   * Without Generics in this declaration, the call site must type cast and can cause the same ClassCastException.
-   * Note that in both cases, the ClassCastException is in the call site, not in this method.
-   * </p>
-   *
-   * @param <T> the object type to be deserialized
-   * @param objectData the serialized object, must not be null
-   * @return the deserialized object
-   * @throws IllegalArgumentException if {@code objectData} is {@code null}
-   * @throws HoodieSerializationException (runtime) if the serialization fails
-   */
-  public static <T> T deserialize(final byte[] objectData) {
-    if (objectData == null) {
-      throw new IllegalArgumentException("The byte[] must not be null");
+  // Returns cached kryo serializer instance if available otherwise creates a new copy of it.
+  private static KryoSerializerInstance borrowKryoSerializerInstance() {
+    final KryoSerializerInstance kryoSerializerInstance = SerializationUtils.serializerRef
+        .getAndSet(null);
+    if (kryoSerializerInstance == null) {
+      return new KryoSerializerInstance();
+    }
+    return kryoSerializerInstance;
+  }
+
+  // Adds kryo serializer into local cache
+  private static void releaseSerializerInstance(KryoSerializerInstance kryoSerializerInstance) {
+    SerializationUtils.serializerRef.compareAndSet(null, kryoSerializerInstance);
+  }
+
+  private static class KryoSerializerInstance implements Serializable {
+
+    private final Kryo kryo;
+    // Caching ByteArrayOutputStream to avoid recreating it for every operation
+    private final ByteArrayOutputStream baos;
+
+    KryoSerializerInstance() {
+      EmptyScalaKryoInstantiator kryoInstantiator = new EmptyScalaKryoInstantiator();
+      kryo = kryoInstantiator.newKryo();
+      baos = new ByteArrayOutputStream(1024);
+      kryo.setRegistrationRequired(false);
+      kryo.register(ObjectWrapper.class);
+    }
+
+    byte[] serialize(Object obj) throws IOException {
+      kryo.reset();
+      baos.reset();
+      Output output = new Output(baos);
+      ObjectWrapper wrapper = new ObjectWrapper(obj);
```

Comment:
For every serialization, a new Wrapper object is created ? I see that you need a class to read/serialize from kryo but anyway to avoid this ?

Reply:
Unfortunately not. I also wanted to avoid it but could not figure out a way without this.

This is also needed for handling null values. By using Wrapper class we can actually serde "null" values.

Reply:
How is this done in spark? does it allocate a wrapper object as well?